### PR TITLE
change grid-with-cat-option-combos column title catOptionCombo form name

### PR DESCRIPTION
**Pull request by deeonwuli**
### :pushpin: References

-   **Issue:** Closes https://app.clickup.com/t/86944kn4z

### :art: Screenshots
Screenshot 2024-03-22 at 01 13 12

### :fire: Notes to the tester

```bash
REACT_APP_DHIS2_BASE_URL=https://dev.eyeseetea.com/who-dev-238/
```

### CODE REVIEW

File: src/data/common/Dhis2DataElement.ts [hunk 1]
```diff
@@ -48,16 +48,20 @@ const dataElementFields = {
             name: true,
             categoryOptions: {
                 id: true,
+                code: true,
                 name: true,
                 shortName: true,
-                formName: true,
+                displayFormName: true,
```
Comment (tokland) [inline src/data/common/Dhis2DataElement.ts:54]: I tried to remove some fields here and still type-checked but there was a runtime error. That should not happened, so I searched for `as` castings. We have this `categoryOptionCombos: getCocOrdered(dataElement.categoryCombo as D2CategoryCombo, config)`. We should never do this kind of casting. It must be removed and the function defined as `function getCocOrdered(categoryCombo: D2DataElement["categoryCombo"]`. Now we can safely remove field; if unused, no TS error should appear.

File: src/data/common/Dhis2DataElement.ts [hunk 2]
```diff
@@ -48,16 +48,20 @@ const dataElementFields = {
...
         },
         categoryOptionCombos: {
             id: true,
             name: true,
             shortName: true,
-            categoryOptions: true,
+            categoryOptions: {
+                id: true,
+                code: true,
+            },
```
Comment (tokland) [inline src/data/common/Dhis2DataElement.ts:64]: Do we need this? after the previous fix in the type, it shows no error, so I guess it's not used. (the form loads normally)

File: src/data/common/Dhis2DataElement.ts [hunk 3]
```diff
@@ -83,7 +87,11 @@ function makeCocOrderArray(namesArray: string[][]): string[] {
 function getCocOrdered(categoryCombo: D2CategoryCombo, config: Dhis2DataStoreDataForm) {
     const allCategoryOptions = categoryCombo.categories
         .map(c => {
-            return c.categoryOptions.flatMap(co => ({ name: co.name, shortName: co.shortName, formName: co.formName }));
+            return c.categoryOptions.flatMap(co => ({
+                name: co.name,
+                shortName: co.shortName,
+                formName: co.displayFormName,
```
Comment (tokland) [inline src/data/common/Dhis2DataElement.ts:93]: should we use also the translated fields for name and shortName?

File: src/domain/common/entities/DataElement.ts [hunk 4]
```diff
@@ -65,7 +65,12 @@ type CategoryCombos = {
     categoryOptionCombos: {
         id: Id;
         name: string;
+        formName: string | undefined;
         shortName: string | undefined;
+        categoryOptions?: {
+            id: Id;
+            code: string;
+        }[];
```
Comment (tokland) [inline src/domain/common/entities/DataElement.ts:73]: Not sure why we need these `categoryOptions?`. It's used in some view model, but where is being filled? If I remove the property and the usage, TS does not complain, so this should mean nobody is filling it.

**Review by tokland — COMMENTED** [review]
Some comments not specifically about Eduardo's new code, but things I've detected while checking the feature. It may be tackled in another PR, if this one must go through.

File: src/domain/common/entities/DataElement.ts (hunk 4 above)
Comment (tokland) [reply src/domain/common/entities/DataElement.ts:73]: Yes, but where is the property being set?

Comment (deeonwuli) [reply src/domain/common/entities/DataElement.ts:73]: it's filled in `Dhis2DataElement.ts` line 134

Comment (tokland) [reply src/domain/common/entities/DataElement.ts:73]: Can you paste the link? at https://github.com/EyeSeeTea/d2-autogen-forms/blob/feat/change-grid-with-cat-option-combos-colTitle-86944kn4z/src/data/common/Dhis2DataElement.ts#L134 I don't see the set.

File: src/data/common/Dhis2DataElement.ts (hunk 3 above)
Comment (deeonwuli) [reply src/data/common/Dhis2DataElement.ts:93]: we already have customizations in the autogen-config to display either formName or shortName (for the grid-with-combos) viewtype

File: src/domain/common/entities/DataElement.ts (hunk 4 above)
Comment (tokland) [reply src/domain/common/entities/DataElement.ts:73]: > dumebi.  okay so it’s set on line 136: const categoryOptionCombos = dataElement.categoryCombo.categoryOptionCombos; the value of categoryOptions is gotten directly from the d2-response`

That's the point, thanks.

The general rule here is not to use optional properties unless strictly necessary.

**Review by MiquelAdell — APPROVED** [review]
I'm approving the PR as I need to generate a release.

Please @deeonwuli if there are still things pending raise a new PR for those. If not ask @tokland to validate the changes

**Pull request merged**